lib/tokens.js

```diff
@@ -449,6 +467,62 @@ export async function resolveToInternalId({token}) {
   return {internalId};
 }
 
+/**
+ * Sets the `minAssuranceForResolution` on the token batches associated
+ *   `internalId`.
+ *
+ * @param {object} options - Options to use.
+ * @param {string} options.internalId - The internal ID of the token batch.
+ * @param {number} options.minAssuranceForResolution - Minimum level of
+ *   assurance required for token resolution.
+ *
+ * @returns {Promise<boolean>} Returns true if update occurred.
+ */
+export async function setMinAssuranceForResolutionOnTokenBatch({
+  internalId, minAssuranceForResolution
+}) {
+  const query = {
+    'tokenBatch.internalId': internalId,
+    'tokenBatch.remainingTokenCount': {$gt: 0}
+  };
+  const $set = {
+    'meta.updated': Date.now(),
+    'tokenBatch.minAssuranceForResolution': minAssuranceForResolution
+  };
+  const collection = database.collections['tokenization-tokenBatch'];
+  const result = await collection.updateMany(
+    query, {$set}, database.writeOptions);
+
+  // return `true` if the update occurred (existing document found)
+  return result.result.n !== 0;
+}
+
+/**
+ * Invalidates the token batch by setting remaining token count to 0, but leaves
+ *   the record intact so previously resolved tokens are trackable.
+ *
+ * @param {object} options - Options to use.
+ * @param {string} options.internalId - The internal ID of the token batch(es).
+ *
+ * @returns {Promise<boolean>} Returns true if update occurred.
+ */
+export async function invalidateTokenBatch({internalId}) {
+  const query = {
+    'tokenBatch.internalId': internalId,
+    'tokenBatch.remainingTokenCount': {$gt: 0}
+  };
+  const $set = {
+    'meta.updated': Date.now(),
+    'tokenBatch.remainingTokenCount': 0
+  };
```

Comment:
Thanks for pointing this out. I had tested this on another repo and was able to get the desired behavior, however when tested directly it did not provide the expected behavior.

I have updated the `invalidateTokenBatch` to `expireTokenBatch`, and when it is called it set the expiration date of the token batch to the current Date. I have also added the parameter that a tokenBatch must not be expired to the `_getBatch` function, which will return a `NotFoundError`. It is possible to add an `allowExpired` flag, however the comment block prior to the `resolve` function notes a token will not resolve if it is expired, so I do not believe that is necessary.

Reply:
You'll want to check to see what happens if the expiration is set to the current date. Does it trigger mongodb's background worker to remove it from the database? If so -- is that ok for auditing purposes or do the token batches need to stick around? I suspect they need to stick around. It may be that we actually need another flag to mark a batch as invalid/unresolvable vs. expiring it early.

Reply:
I have set a new flag `invalid: true` on tokenBatches that have been invalidated. I also set the `remainingTokenCount` to 0 so no new tokens are created from that batch.

Reply:
Ok, thanks, that sounds good, I'll review again soon as I can.